File: SubExprn_Elimn/CSubExp.java
// Short & Simple version to eliminate common sub-expressions in Three-Address Code
import java.util.*;

public class CSubExp
{
    public static void main(String[] args)
    {
        Scanner sc = new Scanner(System.in);        
        System.out.print("Enter number of Three-Address Statements: ");
		int n = sc.nextInt();
        sc.nextLine();

        String[] code = new String[n];
        String[] lhs = new String[n];
        String[] rhs = new String[n];
        
        System.out.println("Enter statements: [eg: t1=a+b]");

        for(int i=0; i<n; i++)
        {
            code[i]=sc.nextLine();
            String[] parts = code[i].split("=");
            lhs[i] = parts[0];
            rhs[i] = parts[1];
        }

        System.out.println("\nOriginal Three Address Code:");
        for(String s: code)
            System.out.println(s);

        Map<String, String> table = new HashMap<>();
        System.out.println("\nOptimized Three Address Code:");
        for(int i=0; i<n; i++)
        {
            if(table.containsKey(rhs[i]))
            {
                System.out.println(lhs[i] + "=" + table.get(rhs[i]));
            }

            else
            {
                table.put(rhs[i],lhs[i]);
                System.out.println(code[i]);
            }
        }
    }
}

File: SubExprn_Elimn/CSubExp.py
def normalize(expr):
    for op in ['+', '*', '&', '|', '^']:
        if op in expr:
            left, right = expr.split(op, 1)
            return op.join(sorted([left.strip(), right.strip()]))
    return expr.strip()


print("Enter number of Three-Address Statements: ", end="")
n = int(input())

code, lhs, rhs = [], [], []

print("Enter statements: [eg: t1=a+b]")
for _ in range(n):
    line = input()
    code.append(line)
    l, r = line.split("=", 1)
    lhs.append(l.strip())
    rhs.append(r.strip())

print("\nOriginal Three Address Code:")
for s in code:
    print(s)

table = {}
print("\nOptimized Three Address Code:")
for i in range(n):
    key = normalize(rhs[i])
    if key in table:
        print(lhs[i] + "=" + table[key])
    else:
        table[key] = lhs[i]
        print(code[i])